incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -115,6 +121,7 @@ func (r *HierarchicalNamespaceReconciler) syncWithForest(log logr.Logger, inst *
 		// to HierarchyConfig reconciler.
 		log.Info("The self-serve subnamespace does not exist", "namespace", nm)
 		r.syncMissing(log, inst, ns)
```

Comment:
If we simply delete the `syncMissing` function, and just replace it with `inst.Status.State = api.Missing` - here's what I _think_ will happen:

* HNSR creates the namespace with the annotation (because we return true). This only succeeds if the namespace doesn't already exist, so there's no possibility of conflict.
* HCR is invoked on the newly created (child) namespace
  * HCR reads the annotation and updates hc.Parent
  * HCR syncs the forest (we'd move the update to `ns.Owner` here)
  * The normal HCR flow will enqueue HCR on the parent (not further discussed here)
  * HCR also re-enqueues HNSR on this object (we're already doing this)
* HNSR runs again, observes that the namespace exists, and clears the "missing" state

So everything in `syncMissing` _seems_ to be unnecessary (given the new changes). Does that seem reasonable?

Reply:
Also - do we even need to see the forest in this reconciler, or enqueue the HCR at all? Would everything work if we just do this:

* Try to load hnsInst and nsInst, but not hcInst
* hnsInst exists:
  * If nsInst doesn't exist, try to create it and set state=Missing
  * If nsInst exists and its annotation is correct, state=OK. Don't check hcInst.Parent, that's up to HCR to ensure that it works.
  * Otherwise, state=Conflict.
* hnsInst is being deleted (i.e. exists but DeletionTimestamp != zero)
  * If nsInst exists and has the right annotation, delete it.
  * Otherwise, remove finalizers

I'll cover setting HNS_MISSING in another comment below, but I think we can move that to the HCR pretty easily, since HCR already needs to load the full list of HNSes anyway so it knows when it can clear its finalizers.

This is much simpler than what we have now. Is there a reason it won't work?

Reply:
Comment 2 sgtm to simply HNSR. I think the current HCR in this PR is already implemented in the way suggested in comment 1.